File: asif-streetshift-sd/web/src/fixtures/policy-map.ts
import type { PolicyMapFeature, PolicyMapPayload } from '../types'

const areas = ['East Village', 'City Center', 'Columbia', 'Marina', 'Cortez', 'Gaslamp']

const fixtureFeatures: PolicyMapFeature[] = Array.from({ length: 261 }, (_, index) => {
  const column = index % 19
  const row = Math.floor(index / 19)
  const x = -117.174 + column * 0.0012 + (row % 2) * 0.00022
  const y = 32.728 - row * 0.00105
  const individualsBefore = index % 7 === 0 ? 2 : index % 4 === 0 ? 1 : 0
  const individualsDelta = index % 13 === 0 ? 2 : index % 9 === 0 ? -1 : 0
  const tentsBefore = index % 6 === 0 ? 3 : index % 3 === 0 ? 1 : 0
  const tentsDelta = index < 93 ? -Math.min(tentsBefore, 1 + (index % 2)) : index < 171 && index % 5 === 0 ? 1 : 0
  const vehiclesBefore = index % 22 === 0 ? 1 : 0
  const vehiclesDelta = index < 93 ? -vehiclesBefore : index % 47 === 0 ? 1 : 0
  const before = {
    individuals: individualsBefore,
    tentsStructures: tentsBefore,
    vehicles: vehiclesBefore,
    adjusted: individualsBefore + 1.75 * tentsBefore + 2.03 * vehiclesBefore,
  }
  const after = {
    individuals: Math.max(0, individualsBefore + individualsDelta),
    tentsStructures: Math.max(0, tentsBefore + tentsDelta),
    vehicles: Math.max(0, vehiclesBefore + vehiclesDelta),
    adjusted: 0,
  }
  after.adjusted = after.individuals + 1.75 * after.tentsStructures + 2.03 * after.vehicles
  const delta = {
    individuals: after.individuals - before.individuals,
    tentsStructures: after.tentsStructures - before.tentsStructures,
    vehicles: after.vehicles - before.vehicles,
    adjusted: after.adjusted - before.adjusted,
  }
  const directionFor = (value: number) => value < 0 ? 'down' as const : value > 0 ? 'up' as const : 'unchanged' as const
  return {
    type: 'Feature',
    geometry: {
      type: 'Polygon',
      coordinates: [[[x, y], [x + 0.00102, y], [x + 0.00102, y - 0.00087], [x, y - 0.00087], [x, y]]],
    },
    properties: {
      blockId: `fixture-${String(index + 1).padStart(3, '0')}`,
      area: areas[index % areas.length],
      before,
      after,
      delta,
      direction: {
        adjusted: directionFor(delta.adjusted),
        individuals: directionFor(delta.individuals),
        tentsStructures: directionFor(delta.tentsStructures),
        vehicles: directionFor(delta.vehicles),
      },
    },
  }
})

export const policyMapFixture: PolicyMapPayload = {
  title: 'Where did Downtown observations change after the 2023 policy rollout?',
  design: 'descriptive_before_after_balanced_panel',
  comparison: {
    before: { reportMonth: '2023-01-01', countDate: '2023-01-31', label: 'Organizer-labeled count date' },
    after: { reportMonth: '2024-01-01', countDate: '2024-01-25', label: 'Organizer-labeled count date' },
    daysApart: 359,
    panelBlocks: 261,
    formula: 'individuals + 1.75 × tents / structures + 2.03 × vehicles',
    interpretation: 'The map displays after minus before for the same 261 blocks.',
  },
  summary: { before: 1314.49, after: 981.8, delta: -332.69, percentChange: -25.31, blocksDown: 93, blocksUp: 78, blocksUnchanged: 90 },
  componentChanges: [
    { component: 'individuals', label: 'Recorded individuals', before: 425, after: 510, delta: 85, percentChange: 20 },
    { component: 'tentsStructures', label: 'Tent / structure marks', before: 470, after: 258, delta: -212, percentChange: -45.1 },
    { component: 'vehicles', label: 'Vehicle marks', before: 33, after: 10, delta: -23, percentChange: -69.7 },
    { component: 'adjusted', label: 'Weighted observed count', before: 1314.49, after: 981.8, delta: -332.69, percentChange: -25.31 },
  ],
  areaChanges: [
    { area: 'East Village', before: 810, after: 526.71, delta: -283.29, percentChange: -34.97, netSharePercent: 85.2 },
    { area: 'City Center', before: 270.25, after: 197.5, delta: -72.75, percentChange: -26.92, netSharePercent: 21.9 },
    { area: 'Marina', before: 23.06, after: 19, delta: -4.06, percentChange: -17.61, netSharePercent: 1.2 },
    { area: 'Columbia', before: 36.28, after: 33.5, delta: -2.78, percentChange: -7.66, netSharePercent: 0.8 },
    { area: 'Gaslamp', before: 45.5, after: 47, delta: 1.5, percentChange: 3.3, netSharePercent: -0.5 },
    { area: 'Cortez', before: 129.4, after: 158.09, delta: 28.69, percentChange: 22.17, netSharePercent: -8.6 },
  ],
  timeline: [
    { id: 'before', date: '2023-01-31', title: 'Before snapshot', description: 'Organizer-labeled block count date.', kind: 'observation' },
    { id: '20th-b', date: '2023-06-30', title: '20th & B began accepting clients', description: 'First Safe Sleeping site opened with 136 tent spaces.', kind: 'site', sourceLabel: 'City of San Diego', sourceUrl: 'https://www.sandiego.gov/outreach2-article/mayor-gloria-announces-opening-first-safe-sleeping-site-san-diegans-experiencing' },
    { id: 'ordinance', date: '2023-07-31', title: 'Ordinance enforcement began', description: 'The City began its scaled approach after outreach and education.', kind: 'policy', sourceLabel: 'City of San Diego', sourceUrl: 'https://www.sandiego.gov/police/services/neighborhood-policing-division/unsafe-camping' },
    { id: 'o-lot', date: '2023-10-21', title: 'O Lot began intakes', description: 'The second Safe Sleeping site began accepting clients.', kind: 'site', sourceLabel: 'City of San Diego', sourceUrl: 'https://www.sandiego.gov/mayor/mayor-gloria-opens-second-safe-sleeping-site-unsheltered-san-diegans' },
    { id: 'after', date: '2024-01-25', title: 'After snapshot', description: 'Organizer-labeled block count date.', kind: 'observation' },
  ],
  geojson: { type: 'FeatureCollection', features: fixtureFeatures },
  safeSleepingSites: [
    { id: '20th-b', name: '20th & B Safe Sleeping', latitude: 32.71829182, longitude: -117.14614399, address: '2145 Caminito Centro (City yard: 1970 B St)', openedDate: '2023-06-30', openingLabel: 'Began accepting clients June 29/30, 2023', capacityLabel: '136 tent spaces · up to 2 people per tent', outsidePanel: true, distanceToPanelMeters: 254, coordinateNote: 'Official City address point; not the precise tent centroid.', sourceLabel: 'City of San Diego', sourceUrl: 'https://docs.sandiego.gov/council_reso_ordinance/rao2024/R-315452.pdf' },
    { id: 'o-lot', name: 'O Lot Safe Sleeping', latitude: 32.72217185, longitude: -117.1475772, address: '1800 Welch Road, San Diego, CA 92101', openedDate: '2023-10-21', openingLabel: 'Began client intakes October 21, 2023', capacityLabel: 'Up to 400 tents at launch · up to 2 people per tent', outsidePanel: true, distanceToPanelMeters: 595, coordinateNote: 'Official City address point; not the precise tent centroid.', sourceLabel: 'City of San Diego', sourceUrl: 'https://www.sandiego.gov/sites/default/files/nora_o_lot_license_agreement.pdf' },
  ],
  ordinanceOffenseLayer: {
    available: true,
    label: 'Police-recorded ordinance offenses',
    description: 'SDPD NIBRS case records whose code section begins 63.0404, rounded and clustered near the study area.',
    sourceSnapshot: '2026-08-20',
    policyWindow: { start: '2023-07-31', end: '2024-01-25' },
    citywideRecordCount: 29,
    citywideUniqueCases: 29,
    validCoordinateCount: 29,
    within500mOfPanelCount: 7,
    nearbyClusterCount: 6,
    clusters: [
      { id: 'ordinance-cluster-01', latitude: 32.723, longitude: -117.141, recordCount: 2, firstDate: '2023-08-08', lastDate: '2023-12-24', subsection: '63.0404(c)', neighborhood: 'Balboa Park' },
      { id: 'ordinance-cluster-02', latitude: 32.720, longitude: -117.140, recordCount: 1, firstDate: '2023-08-09', lastDate: '2023-08-09', subsection: '63.0404(c)', neighborhood: 'Golden Hill' },
      { id: 'ordinance-cluster-03', latitude: 32.718, longitude: -117.149, recordCount: 1, firstDate: '2023-08-21', lastDate: '2023-08-21', subsection: '63.0404(c)', neighborhood: 'East Village' },
      { id: 'ordinance-cluster-04', latitude: 32.725, longitude: -117.158, recordCount: 1, firstDate: '2023-11-03', lastDate: '2023-11-03', subsection: '63.0404(c)', neighborhood: 'Balboa Park' },
      { id: 'ordinance-cluster-05', latitude: 32.702, longitude: -117.144, recordCount: 1, firstDate: '2023-12-31', lastDate: '2023-12-31', subsection: '63.0404(a)', neighborhood: 'Barrio Logan' },
      { id: 'ordinance-cluster-06', latitude: 32.726, longitude: -117.158, recordCount: 1, firstDate: '2024-01-06', lastDate: '2024-01-06', subsection: '63.0404(c)', neighborhood: 'Balboa Park' },
    ],
    caveat: 'These are not raid, citation, arrest, outreach, or encampment-removal locations. One record does not necessarily represent one person.',
    sourceLabel: 'City of San Diego — Police NIBRS Crime Offenses',
    sourceUrl: 'https://data.sandiego.gov/datasets/police-nibrs/',
  },
  enforcementLayer: {
    available: false,
    title: 'Event-level enforcement locations are not public',
    explanation: 'The City Auditor analyzed police records supplied internally, but no public row-level dataset was found. Get It Done points are resident reports—not raids, citations, or confirmed abatements.',
    alternativeLabel: 'Public encampment reports are a separate proxy and are not plotted as enforcement.',
    sourceLabel: 'City Auditor, 2026',
    sourceUrl: 'https://www.sandiego.gov/sites/default/files/2026-04/performance-audit-of-the-city-s-response-to-homeless-encampments-since-the-unsafe-camping-ordinance.pdf',
  },
  auditorFindings: [
    { id: 'movement', metric: 'Not traceable', title: 'The City could not determine where people moved', detail: 'The official audit says available data cannot establish specific geographic movement within the City.', sourceLabel: 'City Auditor, 2026', sourceUrl: 'https://www.sandiego.gov/sites/default/files/2026-04/performance-audit-of-the-city-s-response-to-homeless-encampments-since-the-unsafe-camping-ordinance.pdf' },
    { id: 'reports', metric: '+45%', title: 'Downtown encampment reports increased', detail: 'Get It Done encampment reports rose by 6,938 in the two years after versus the two years before the ordinance.', sourceLabel: 'City Auditor, 2026', sourceUrl: 'https://www.sandiego.gov/sites/default/files/2026-04/performance-audit-of-the-city-s-response-to-homeless-encampments-since-the-unsafe-camping-ordinance.pdf' },
  ],
  sources: [
    { id: 'organizer', label: 'Building for Good organizer block dataset', url: 'https://drive.google.com/drive/folders/1cJ6_sIiJ8FG_IqZ7LN4ET__ZR_N8yWwv' },
    { id: 'audit', label: 'City Auditor performance audit', url: 'https://www.sandiego.gov/sites/default/files/2026-04/performance-audit-of-the-city-s-response-to-homeless-encampments-since-the-unsafe-camping-ordinance.pdf' },
  ],
  caveat: 'This is a descriptive before/after comparison. It does not identify people, trace movement, or isolate the effect of any one policy.',
}

File: asif-streetshift-sd/backend/policy_map.py
from __future__ import annotations

import json
from functools import lru_cache
from typing import Any

import pandas as pd

from .config import DATA_DIR, ROOT_DIR


BEFORE_REPORT_MONTH = pd.Timestamp("2023-01-01")
AFTER_REPORT_MONTH = pd.Timestamp("2024-01-01")
BEFORE_COUNT_DATE = pd.Timestamp("2023-01-31")
AFTER_COUNT_DATE = pd.Timestamp("2024-01-25")
PANEL_BLOCKS = 261
TENT_MULTIPLIER = 1.75
VEHICLE_MULTIPLIER = 2.03
FORMULA = "individuals + 1.75 × tents / structures + 2.03 × vehicles"

AUDITOR_REPORT_URL = (
    "https://www.sandiego.gov/sites/default/files/2026-04/"
    "performance-audit-of-the-city-s-response-to-homeless-encampments-"
    "since-the-unsafe-camping-ordinance.pdf"
)
FIRST_SITE_URL = (
    "https://www.sandiego.gov/outreach2-article/"
    "mayor-gloria-announces-opening-first-safe-sleeping-site-san-diegans-experiencing"
)
SECOND_SITE_URL = (
    "https://www.sandiego.gov/mayor/"
    "mayor-gloria-opens-second-safe-sleeping-site-unsheltered-san-diegans"
)
ORDINANCE_URL = (
    "https://www.sandiego.gov/police/services/neighborhood-policing-division/"
    "unsafe-camping"
)
GET_IT_DONE_URL = "https://data.sandiego.gov/datasets/get-it-done-reports/"
GET_IT_DONE_GUIDANCE_URL = (
    "https://data.sandiego.gov/help/articles/tips-getitdone-311-requests/"
)
NIBRS_DATASET_URL = "https://data.sandiego.gov/datasets/police-nibrs/"
NIBRS_2023_URL = "https://seshat.datasd.org/police_nibrs/pd_nibrs_2023_datasd.csv"
NIBRS_2024_URL = "https://seshat.datasd.org/police_nibrs/pd_nibrs_2024_datasd.csv"

COMPONENTS = (
    ("individuals", "individuals", "Individuals"),
    ("tents_structures", "tentsStructures", "Tents / structures"),
    ("vehicles", "vehicles", "Vehicles"),
)


def _number(value: float | int, digits: int = 2) -> float | int:
    """Return JSON-friendly numbers while retaining meaningful decimals."""
    rounded = round(float(value), digits)
    return int(rounded) if rounded.is_integer() else rounded


def _percent_change(before: float, delta: float, digits: int = 2) -> float | None:
    if before == 0:
        return None
    return round(delta / before * 100, digits)


def _direction(delta: float) -> str:
    if delta < 0:
        return "down"
    if delta > 0:
        return "up"
    return "unchanged"


def _validate_period(
    period: pd.DataFrame,
    report_month: pd.Timestamp,
    expected_count_date: pd.Timestamp,
) -> None:
    if len(period) != PANEL_BLOCKS or period["block_id"].nunique() != PANEL_BLOCKS:
        raise ValueError(
            f"{report_month:%Y-%m} must contain {PANEL_BLOCKS} unique panel blocks"
        )
    count_dates = set(period["count_date"].dropna().unique())
    if count_dates != {expected_count_date}:
        formatted = sorted(pd.Timestamp(date).strftime("%Y-%m-%d") for date in count_dates)
        raise ValueError(
            f"{report_month:%Y-%m} should use organizer count date "
            f"{expected_count_date:%Y-%m-%d}; found {formatted}"
        )


def _safe_sleeping_sites() -> list[dict[str, Any]]:
    return [
        {
            "id": "safe-sleeping-20th-b",
            "name": "20th & B Safe Sleeping Site",
            "latitude": 32.71829182,
            "longitude": -117.14614399,
            "address": "2145 Caminito Centro, San Diego, CA 92102",
            "facilityAddress": "1970 B Street, San Diego, CA 92102",
            "openedDate": "2023-06-30",
            "openingLabel": "Began accepting clients June 29–30, 2023",
            "openedDateNote": (
                "City materials place initial client arrivals across June 29–30, 2023; "
                "June 30 is used on this timeline."
            ),
            "tentSpaces": 136,
            "peoplePerTentMax": 2,
            "capacityLabel": "136 tent spaces · up to 2 people per tent",
            "insideCountPanel": False,
            "outsidePanel": True,
            "distanceToPanelMeters": 254,
            "coordinateNote": (
                "Official City address point; it does not claim to be the precise tent-area centroid."
            ),
            "sourceLabel": "City of San Diego — first Safe Sleeping site announcement",
            "sourceUrl": FIRST_SITE_URL,
            "addressSourceUrl": (
                "https://docs.sandiego.gov/council_reso_ordinance/rao2024/R-315452.pdf"
            ),
        },
        {
            "id": "safe-sleeping-o-lot",
            "name": "O Lot Safe Sleeping Site",
            "latitude": 32.72217185,
            "longitude": -117.14757720,
            "address": "1800 Welch Road, San Diego, CA 92101",
            "facilityAddress": None,
            "openedDate": "2023-10-21",
            "openingLabel": "Began client intakes October 21, 2023",
            "openedDateNote": "The City announced that client intakes would begin October 21, 2023.",
            "tentSpaces": 400,
            "peoplePerTentMax": 2,
            "capacityLabel": "Up to 400 tents at launch · up to 2 people per tent",
            "insideCountPanel": False,
            "outsidePanel": True,
            "distanceToPanelMeters": 595,
            "coordinateNote": (
                "Official City address point; it does not claim to be the precise tent-area centroid."
            ),
            "sourceLabel": "City of San Diego — second Safe Sleeping site announcement",
            "sourceUrl": SECOND_SITE_URL,
            "addressSourceUrl": (
                "https://docs.sandiego.gov/council_reso_ordinance/rao2024/R-315812.pdf"
            ),
        },
    ]


def _ordinance_offense_layer() -> dict[str, Any]:
    """Return a privacy-reduced snapshot of public NIBRS ordinance records."""
    with (ROOT_DIR / "data" / "derived" / "ordinance_offense_clusters.json").open(
        encoding="utf-8"
    ) as source:
        snapshot = json.load(source)
    return {
        "available": True,
        "label": "Police-recorded ordinance offenses",
        "description": (
            "SDPD NIBRS case records whose code_section begins 63.0404. Markers are "
            "rounded, clustered locations within 500 meters of the study area."
        ),
        **snapshot,
        "caveat": (
            "These are not raid, citation, arrest, outreach, or encampment-removal "
            "locations. They are offenses recorded in police case reports; one record "
            "does not necessarily represent one person, and encounters without a NIBRS "
            "offense are absent."
        ),
        "sourceLabel": "City of San Diego — Police NIBRS Crime Offenses",
        "sourceUrl": NIBRS_DATASET_URL,
        "sourceFiles": [NIBRS_2023_URL, NIBRS_2024_URL],
    }


def build_policy_map_payload(
    panel: pd.DataFrame | None = None,
    grid: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Build the focused, evidence-bounded policy-period map contract.

    The endpoint compares the same 261 organizer-defined blocks on two count
    nights. It describes changes in observations; it does not track people or
    estimate that any policy caused those changes.
    """
    data = (
        pd.read_csv(
            DATA_DIR / "BlockLevel_Counts_Panel261.csv",
            parse_dates=["report_month", "count_date"],
        )
        if panel is None
        else panel.copy()
    )
    for column in ("report_month", "count_date"):
        data[column] = pd.to_datetime(data[column], errors="raise")
    for column in ("individuals", "tents_structures", "vehicles"):
        data[column] = pd.to_numeric(data[column], errors="raise")
    data["adjusted"] = (
        data["individuals"]
        + TENT_MULTIPLIER * data["tents_structures"]
        + VEHICLE_MULTIPLIER * data["vehicles"]
    )

    before = data[data["report_month"].eq(BEFORE_REPORT_MONTH)].copy()
    after = data[data["report_month"].eq(AFTER_REPORT_MONTH)].copy()
    _validate_period(before, BEFORE_REPORT_MONTH, BEFORE_COUNT_DATE)
    _validate_period(after, AFTER_REPORT_MONTH, AFTER_COUNT_DATE)

    raw_columns = [
        "block_id",
        "area",
        "individuals",
        "tents_structures",
        "vehicles",
        "adjusted",
    ]
    joined = before[raw_columns].merge(
        after[raw_columns],
        on="block_id",
        suffixes=("_before", "_after"),
        how="inner",
        validate="one_to_one",
    )
    if len(joined) != PANEL_BLOCKS:
        raise ValueError(
            f"Balanced panel should contain {PANEL_BLOCKS} joined blocks; found {len(joined)}"
        )
    if not joined["area_before"].eq(joined["area_after"]).all():
        raise ValueError("A balanced-panel block changed area labels between count dates")

    if grid is None:
        with (DATA_DIR / "Downtown_BlockGrid.geojson").open(encoding="utf-8") as source:
            grid = json.load(source)
    geometries = {
        feature["properties"]["block_id"]: feature["geometry"]
        for feature in grid["features"]
    }
    missing_geometry = set(joined["block_id"]) - set(geometries)
    if missing_geometry:
        raise ValueError(
            f"Block grid is missing {len(missing_geometry)} balanced-panel geometries"
        )

    features: list[dict[str, Any]] = []
    for row in joined.sort_values("block_id").itertuples(index=False):
        before_values = {
            "individuals": _number(row.individuals_before, 0),
            "tentsStructures": _number(row.tents_structures_before, 0),
            "vehicles": _number(row.vehicles_before, 0),
            "adjusted": _number(row.adjusted_before, 2),
        }
        after_values = {
            "individuals": _number(row.individuals_after, 0),
            "tentsStructures": _number(row.tents_structures_after, 0),
            "vehicles": _number(row.vehicles_after, 0),
            "adjusted": _number(row.adjusted_after, 2),
        }
        delta_values = {
            key: _number(float(after_values[key]) - float(before_values[key]), 2)
            for key in before_values
        }
        direction = {key: _direction(float(value)) for key, value in delta_values.items()}
        features.append(
            {
                "type": "Feature",
                "geometry": geometries[row.block_id],
                "properties": {
                    "blockId": row.block_id,
                    "area": row.area_before,
                    "before": before_values,
                    "after": after_values,
                    "delta": delta_values,
                    "direction": direction,
                },
            }
        )

    before_total = float(before["adjusted"].sum())
    after_total = float(after["adjusted"].sum())
    total_delta = after_total - before_total

    component_changes = []
    for source_column, component, label in COMPONENTS:
        component_before = float(before[source_column].sum())
        component_after = float(after[source_column].sum())
        component_delta = component_after - component_before
        component_changes.append(
            {
                "component": component,
                "label": label,
                "before": _number(component_before, 0),
                "after": _number(component_after, 0),
                "delta": _number(component_delta, 0),
                "percentChange": _percent_change(component_before, component_delta, 1),
                "direction": _direction(component_delta),
            }
        )

    area_changes = []
    for area, rows in joined.groupby("area_before"):
        area_before = float(rows["adjusted_before"].sum())
        area_after = float(rows["adjusted_after"].sum())
        area_delta = area_after - area_before
        area_changes.append(
            {
                "area": area,
                "blocks": int(len(rows)),
                "before": _number(area_before, 2),
                "after": _number(area_after, 2),
                "delta": _number(area_delta, 2),
                "percentChange": _percent_change(area_before, area_delta),
                "direction": _direction(area_delta),
            }
        )
    area_changes.sort(key=lambda item: float(item["delta"]))

    events = [
        {
            "id": "before-count",
            "kind": "observation",
            "date": "2023-01-31",
            "title": "Before count",
            "description": "Organizer-labeled overnight count across the fixed 261-block panel.",
            "sourceLabel": "Building for Good organizer dataset",
            "sourceUrl": (
                "https://github.com/sandiegodata-projects/downtown-partnership"
            ),
        },
        {
            "id": "safe-sleeping-20th-b",
            "kind": "site",
            "date": "2023-06-30",
            "title": "20th & B Safe Sleeping site opened",
            "description": "First City Safe Sleeping site began serving clients.",
            "sourceLabel": "City of San Diego — opening announcement",
            "sourceUrl": FIRST_SITE_URL,
        },
        {
            "id": "unsafe-camping-enforcement",
            "kind": "policy",
            "date": "2023-07-31",
            "title": "Unsafe Camping Ordinance enforcement began",
            "description": (
                "The City began its progressive enforcement approach after outreach "
                "and education."
            ),
            "sourceLabel": "City of San Diego — Unsafe Camping Ordinance",
            "sourceUrl": ORDINANCE_URL,
        },
        {
            "id": "safe-sleeping-o-lot",
            "kind": "site",
            "date": "2023-10-21",
            "title": "O Lot Safe Sleeping site began intakes",
            "description": "The second City Safe Sleeping site began client intakes.",
            "sourceLabel": "City of San Diego — opening announcement",
            "sourceUrl": SECOND_SITE_URL,
        },
        {
            "id": "after-count",
            "kind": "observation",
            "date": "2024-01-25",
            "title": "After count",
            "description": "Organizer-labeled overnight count across the same 261 blocks.",
            "sourceLabel": "Building for Good organizer dataset",
            "sourceUrl": (
                "https://github.com/sandiegodata-projects/downtown-partnership"
            ),
        },
    ]

    return {
        "title": "Where did downtown street observations change after the 2023 policy rollout?",
        "design": "descriptive_before_after_balanced_panel",
        "comparison": {
            "before": {
                "reportMonth": "2023-01-01",
                "countDate": "2023-01-31",
                "label": "Overnight count conducted January 31, 2023",
            },
            "after": {
                "reportMonth": "2024-01-01",
                "countDate": "2024-01-25",
                "label": "Overnight count conducted January 25, 2024",
            },
            "monthsApart": 12,
            "daysApart": int((AFTER_COUNT_DATE - BEFORE_COUNT_DATE).days),
            "panelBlocks": PANEL_BLOCKS,
            "formula": FORMULA,
            "unit": "adjusted observed count",
            "interpretation": (
                "Every color is the January 25, 2024 observation minus the "
                "January 31, 2023 observation for the same block."
            ),
        },
        "summary": {
            "before": _number(before_total, 2),
            "after": _number(after_total, 2),
            "delta": _number(total_delta, 2),
            "percentChange": _percent_change(before_total, total_delta),
            "blocksDown": sum(
                feature["properties"]["direction"]["adjusted"] == "down"
                for feature in features
            ),
            "blocksUp": sum(
                feature["properties"]["direction"]["adjusted"] == "up"
                for feature in features
            ),
            "blocksUnchanged": sum(
                feature["properties"]["direction"]["adjusted"] == "unchanged"
                for feature in features
            ),
        },
        "componentChanges": component_changes,
        "componentInterpretation": (
            "The adjusted observation fell because recorded tents / structures and "
            "vehicles declined, while directly recorded individuals rose from 425 to "
            "510. The map can separate these components instead of presenting the "
            "weighted total as a literal population count."
        ),
        "areaChanges": area_changes,
        "timeline": events,
        "safeSleepingSites": _safe_sleeping_sites(),
        "ordinanceOffenseLayer": _ordinance_offense_layer(),
        "enforcementLayer": {
            "available": False,
            "label": "Event-level enforcement locations are not publicly available",
            "title": "Cleanup and abatement locations are not publicly available",
            "explanation": (
                "The official public Get It Done data contains geolocated resident "
                "reports, not verified enforcement contacts, citations, arrests, raids, "
                "or abatements. Plotting those reports as enforcement would mislabel the "
                "evidence, so this layer is intentionally disabled."
            ),
            "alternativeLabel": (
                "Resident reports are a separate proxy and are not plotted as enforcement."
            ),
            "publicProxy": {
                "name": "Get It Done encampment reports",
                "appropriateLabel": "Resident-reported encampment locations",
                "notEquivalentTo": "Enforcement or abatement locations",
                "datasetUrl": GET_IT_DONE_URL,
                "guidanceUrl": GET_IT_DONE_GUIDANCE_URL,
            },
            "sourceLabel": "City Auditor — Performance Audit (April 2026)",
            "sourceUrl": AUDITOR_REPORT_URL,
        },
        "auditorFindings": [
            {
                "id": "movement-not-identifiable",
                "metric": "Movement unknown",
                "title": "The City could not determine where people moved",
                "detail": (
                    "The City Auditor was unable to determine the specific geographic "
                    "movement of the homeless population. Block changes show where "
                    "observations differed, not where any individual person went."
                ),
                "sourceLabel": "City Auditor — Performance Audit (April 2026)",
                "sourceUrl": AUDITOR_REPORT_URL,
            },
            {
                "id": "downtown-reports-increased",
                "metric": "+45%",
                "title": "Downtown encampment reports increased",
                "detail": (
                    "Downtown received 6,938 more Get It Done encampment reports in the "
                    "two years after the ordinance than in the two years before it, a 45% "
                    "increase. Fewer weighted observations therefore do not establish "
                    "fewer resident-reported encampment concerns."
                ),
                "sourceLabel": "City Auditor — Performance Audit (April 2026)",
                "sourceUrl": AUDITOR_REPORT_URL,
            },
        ],
        "geojson": {"type": "FeatureCollection", "features": features},
        "caveat": (
            "This is a descriptive comparison of two single-night observations, not a "
            "causal estimate or a population-movement tracker. The Safe Sleeping sites "
            "and ordinance implementation overlap in time, there is no untreated control "
            "group, and other conditions may explain some or all of the change. Both site "
            "markers are outside the fixed downtown count panel."
        ),
        "sources": [
            {
                "id": "organizer",
                "label": "Organizer block-count repository",
                "url": "https://github.com/sandiegodata-projects/downtown-partnership",
            },
            {"id": "first-site", "label": "City — first Safe Sleeping site", "url": FIRST_SITE_URL},
            {"id": "second-site", "label": "City — second Safe Sleeping site", "url": SECOND_SITE_URL},
            {"id": "ordinance", "label": "City — Unsafe Camping Ordinance", "url": ORDINANCE_URL},
            {"id": "audit", "label": "City Auditor — 2026 performance audit", "url": AUDITOR_REPORT_URL},
            {"id": "get-it-done", "label": "City Open Data — Get It Done reports", "url": GET_IT_DONE_URL},
            {"id": "nibrs", "label": "City Open Data — Police NIBRS offenses", "url": NIBRS_DATASET_URL},
        ],
    }


@lru_cache(maxsize=1)
def policy_map_payload() -> dict[str, Any]:
    return build_policy_map_payload()
